File: Java/com/leetcode/bestTimeToBuyAndSellStock/BestTimeToBuyAndSellStock.java
package com.leetcode.bestTimeToBuyAndSellStock;

public class BestTimeToBuyAndSellStock {
    public int maxProfit(int[] prices) {
        // If only permitted to complete at most one transaction, we must find an ascending range that the min on the left and max on the right, and also has largest difference.
        if (prices == null || prices.length == 0) {
            return 0;
        }
        int min = Integer.MAX_VALUE;
        int diff = Integer.MIN_VALUE;
        
        for (int i = 0; i < prices.length; ++i) {
            if (prices[i] < min) {
                min = prices[i];
            }
            if (prices[i] - min > diff) {
                diff = prices[i] - min;
            }
        }
        
        return diff;
    }
}

File: Python/best_time_to_buy_and_sell_stock/best_time_to_buy_and_sell_stock.py
class Solution(object):
	def maxProfit(self, prices):
		if len(prices) < 1:
			return 0
		pmin = prices[0]
		max_profit = 0
		for i in range(0, len(prices)):
			pmin = min(pmin, prices[i])
			max_profit = max(max_profit, prices[i] - pmin)	
		return max_profit
